sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -132,6 +132,7 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate
   }
 
   override def children: Seq[Expression] = value +: list
+  lazy val inSetConvertible = children.forall(_.deterministic)
```

Comment:
What you want is just `children.forall(_.isInstanceOf[Literal])`, isn't it?
It's because all literals are deterministic.